### PR TITLE
Extending the DCGAN example implemented by gluon API to provide a more straight-forward evaluation on the generated image

### PR DESCRIPTION
**Pull request by pengxin99**
## Description ##
On using the example script “example\gluon\dcgan.py”,  it is noticed the loss and metrics of netG and netD are not straight-forward to evaluate the quality of the generated images. (the absolute value of the metrics/loss fluctuate outstandingly, and there is no stable trends can be observed).

According to the paper [Unsupervised Representation Learning with Deep Convolutional Generative Adversarial Networks](https://arxiv.org/abs/1511.06434) and a tensorflow implementation [openai/improved-gan](https://github.com/openai/improved-gan) “inception score” is a methedodoly to evaluate the image quality generate by DCGAN. Via this PR, I implemented the inception score for DCGAN by using the pre-trained CNN model provided by gluon model zoo. The test result is expected.

reference to #12612 
@pengzhao-intel @juliusshufan

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Add inception_score.py as inception_score calculation module
- [x] Modify dcgan.py to make it easy to call each functions, including the inception_score function
- [x] Add the results(generated pics and inception_score curve chart)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by juliusshufan** [comment]
@sandeep-krishnamurthy  Thanks for add the label, will you keep on reviewing this PR?

File: example/gluon/DCgan/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# DCgan in MXNet
```
Comment (sandeep-krishnamurthy) [inline example/gluon/DCgan/README.md]: May be we should call it DCGAN everywhere.

File: example/gluon/DCgan/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+train the dcgan with mxnet, and eval dcgan model with inception_score.
```
Comment (sandeep-krishnamurthy) [inline example/gluon/DCgan/README.md]: [Deep Convolutional Generative Adversarial Network (DCGAN)](https://arxiv.org/abs/1511.06434) implementation with Apache MXNet GLUON. This implementation uses, [inception_score](https://github.com/openai/improved-gan) to evaluate the model.

You can use this reference implementation on MNIST and CIFAR10 dataset.

File: example/gluon/DCgan/dcgan.py [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+from mxnet import autograd
+import numpy as np
+import logging
+from datetime import datetime
+import os
+import time
+
+from inception_score import get_inception_score
+
+def fill_buf(buf, i, img, shape):
```
Comment (sandeep-krishnamurthy) [inline example/gluon/DCgan/dcgan.py]: Please add code docs for easy readability and maintaining this script.

File: example/gluon/DCgan/dcgan.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+                                             'and then use inception_score to metric the result.')
+parser.add_argument('--dataset', type=str, default='cifar10', help='dataset to use. options are cifar10 and mnist.')
+parser.add_argument('--batch-size', type=int, default=64, help='input batch size')
+parser.add_argument('--nz', type=int, default=100, help='size of the latent z vector')
+parser.add_argument('--ngf', type=int, default=64, help='the channel of each generator filter layer.')
+parser.add_argument('--ndf', type=int, default=64, help='the channel of each descriminator filter layer.')
+parser.add_argument('--nepoch', type=int, default=25, help='number of epochs to train for')
+parser.add_argument('--lr', type=float, default=0.0002, help='learning rate, default=0.0002')
+parser.add_argument('--beta1', type=float, default=0.5, help='beta1 for adam. default=0.5')
+parser.add_argument('--cuda', action='store_true', help='enables cuda')
```
Comment (sandeep-krishnamurthy) [inline example/gluon/DCgan/dcgan.py]: why is this required? if --ngpu is given

File: example/gluon/DCgan/dcgan.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+print(opt)
+
+logging.basicConfig(level=logging.DEBUG)
+ngpu = int(opt.ngpu)
+nz = int(opt.nz)
+ngf = int(opt.ngf)
+ndf = int(opt.ndf)
+nc = 3
+if opt.cuda:
+    ctx = mx.gpu(0)
```
Comment (sandeep-krishnamurthy) [inline example/gluon/DCgan/dcgan.py]: not using ngpu?

File: example/gluon/DCgan/dcgan.py [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+            ############################
+            # (1) Update D network: maximize log(D(x)) + log(1 - D(G(z)))
+            ###########################
+            # train with real_t
+            data = data.as_in_context(ctx)
+            noise = mx.nd.random.normal(0, 1, shape=(opt.batch_size, nz, 1, 1), ctx=ctx)
+
+            with autograd.record():
+                output = netD(data)
+                output = output.reshape((opt.batch_size, 2))
```
Comment (sandeep-krishnamurthy) [inline example/gluon/DCgan/dcgan.py]: why is this required?

File: example/gluon/DCgan/dcgan.py [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+                errD_real = loss(output, real_label)
+                metric.update([real_label, ], [output, ])
```
Comment (sandeep-krishnamurthy) [inline example/gluon/DCgan/dcgan.py]: Please move metric update and backward pass out of autograd.record()

File: example/gluon/DCgan/dcgan.py [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+            trainerD.step(opt.batch_size)
+
+            ############################
+            # (2) Update G network: maximize log(D(G(z)))
+            ###########################
+            with autograd.record():
+                output = netD(fake)
+                output = output.reshape((-1, 2))
+                errG = loss(output, real_label)
+                errG.backward()
```
Comment (sandeep-krishnamurthy) [inline example/gluon/DCgan/dcgan.py]: Move backward pass out of autograd.record

File: example/gluon/DCgan/dcgan.py [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+
+            trainerG.step(opt.batch_size)
+
+            name, acc = metric.get()
+            # logging.info('speed: {} samples/s'.format(opt.batch_size / (time.time() - btic)))
```
Comment (sandeep-krishnamurthy) [inline example/gluon/DCgan/dcgan.py]: remove?

File: example/gluon/DCgan/dcgan.py [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+            logging.info('discriminator loss = %f, generator loss = %f, binary training acc = %f at iter %d epoch %d'
+                         % (mx.nd.mean(errD).asscalar(), mx.nd.mean(errG).asscalar(), acc, iter, epoch))
+            if iter % 100 == 0:
```
Comment (sandeep-krishnamurthy) [inline example/gluon/DCgan/dcgan.py]: default iter is 25. why % 100 here?

File: example/gluon/DCgan/inception_score.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+import numpy as np
+import math
+import sys
+
+
+inception_model = None
+
+
+def get_inception_score(images, splits=10):
+
```
Comment (sandeep-krishnamurthy) [inline example/gluon/DCgan/inception_score.py]: Please add more description and code comments.

File: example/gluon/DCgan/dcgan.py (hunk 10 above)
Comment (pengxin99) [reply example/gluon/DCgan/dcgan.py]: this '100' mean every per 100 iters, save the generated images and inception_score, i have set it as a parameter. default iter was not set before, and default epoch is 25.

**Comment by pengxin99** [comment]
@sandeep-krishnamurthy Thanks for your review. I have fixed them accordingly. :)

**Comment by sandeep-krishnamurthy** [comment]
> @sandeep-krishnamurthy Thanks for your review. I have fixed them accordingly. :)

Thank you. Can we please answer the following?
> Maybe should move this under "examples/gluon/gan/dcgan" so we can have all gan examples under gan.
> Can you check if from any docs we are referring to the example/gluon/dcgan.py that are you moving, so we don't have broken links.

@lebeg - Are examples part of CI nightly tests? Should this be included under the tests?
@aaronmarkham - Can you please review this example?

File: example/gluon/DCGAN/README.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+[Deep Convolutional Generative Adversarial Networks(DCGAN)](https://arxiv.org/abs/1511.06434) implementation with Apache MXNet GLUON.
+This implementation uses [inception_score](https://github.com/openai/improved-gan) to evaluate the model.
+
+You can use this reference implementation on MNIST and CIFAR10 dataset.
+
+
+#### Generated pic(use dataset cifar10)
+![Generated pic](https://github.com/pengxin99/incubator-mxnet/blob/dcgan-inception_score/example/gluon/DCGAN/pic/fake_img_iter_13900.png)
+
+#### Generated pic(use dataset mnist)
```
Comment (aaronmarkham) [inline example/gluon/DCGAN/README.md]:
```suggestion
#### Generated image output examples from the MNIST dataset
```

File: example/gluon/DCGAN/README.md [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+![Generated pic](https://github.com/pengxin99/incubator-mxnet/blob/dcgan-inception_score/example/gluon/DCGAN/pic/fake_img_iter_21700.png)
```
Comment (aaronmarkham) [inline example/gluon/DCGAN/README.md]:
```suggestion
![Generated image output examples from the MNIST dataset](https://github.com/pengxin99/incubator-mxnet/blob/dcgan-inception_score/example/gluon/DCGAN/pic/fake_img_iter_21700.png)
```

File: example/gluon/DCGAN/README.md [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+#### inception_score in cpu and gpu (the real image`s score is around 3.3)
+CPU & GPU
+
+![inception_socre_with_cpu](https://github.com/pengxin99/incubator-mxnet/blob/dcgan-inception_score/example/gluon/DCGAN/pic/inception_score_cifar10_cpu.png)
```
Comment (aaronmarkham) [inline example/gluon/DCGAN/README.md]:
```suggestion
![inception score with CPU](https://github.com/pengxin99/incubator-mxnet/blob/dcgan-inception_score/example/gluon/DCGAN/pic/inception_score_cifar10_cpu.png)
```
When you put in the alt text like this, it becomes readable by screen readers and the like... and it's searchable.

File: example/gluon/DCGAN/README.md [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+![inception_score_with_gpu](https://github.com/pengxin99/incubator-mxnet/blob/dcgan-inception_score/example/gluon/DCGAN/pic/inception_score_cifar10.png)
```
Comment (aaronmarkham) [inline example/gluon/DCGAN/README.md]:
```suggestion
![inception score with GPU](https://github.com/pengxin99/incubator-mxnet/blob/dcgan-inception_score/example/gluon/DCGAN/pic/inception_score_cifar10.png)
```

File: example/gluon/DCGAN/README.md [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+## Quick start
+use below code to see the configurations you can set:
+```python
```
Comment (aaronmarkham) [inline example/gluon/DCGAN/README.md]:
```suggestion
```bash
```

File: example/gluon/DCGAN/README.md [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+![Generated pic](https://github.com/pengxin99/incubator-mxnet/blob/dcgan-inception_score/example/gluon/DCGAN/pic/fake_img_iter_21700.png)
+
+#### inception_score in cpu and gpu (the real image`s score is around 3.3)
+CPU & GPU
+
+![inception_socre_with_cpu](https://github.com/pengxin99/incubator-mxnet/blob/dcgan-inception_score/example/gluon/DCGAN/pic/inception_score_cifar10_cpu.png)
+![inception_score_with_gpu](https://github.com/pengxin99/incubator-mxnet/blob/dcgan-inception_score/example/gluon/DCGAN/pic/inception_score_cifar10.png)
+
+## Quick start
+use below code to see the configurations you can set:
```
Comment (aaronmarkham) [inline example/gluon/DCGAN/README.md]:
```suggestion
Use the following code to see the configurations you can set:
```

File: example/gluon/DCGAN/README.md [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+      --netD NETD           path to netD (to continue training)
+      --outf OUTF           folder to output images and model checkpoints
+      --check-point CHECK_POINT
+                            save results at each epoch or not
+      --inception_score INCEPTION_SCORE
+                            To record the inception_score, default is True.
+
+
+use below code to train DCGAN model with default configurations and dataset(cifar10), and metric with inception_score:
+```python
```
Comment (aaronmarkham) [inline example/gluon/DCGAN/README.md]:
```suggestion
```bash
```

File: example/gluon/DCGAN/README.md [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+      --netG NETG           path to netG (to continue training)
+      --netD NETD           path to netD (to continue training)
+      --outf OUTF           folder to output images and model checkpoints
+      --check-point CHECK_POINT
+                            save results at each epoch or not
+      --inception_score INCEPTION_SCORE
+                            To record the inception_score, default is True.
+
+
+use below code to train DCGAN model with default configurations and dataset(cifar10), and metric with inception_score:
```
Comment (aaronmarkham) [inline example/gluon/DCGAN/README.md]:
```suggestion
Use the following Python script to train a DCGAN model with default configurations using the CIFAR-10 dataset and record metrics with `inception_score`:
```

**Comment by pengxin99** [comment]
> @aaronmarkham Also, you're deleting a file that has links to it elsewhere in the repo, specifically /example/gluon/sn_gan. Make sure you deal with that. Delete the links or update as appropriate...
Thanks for your suggestions, I only found that there are relative link in the comment section of these files(/example/gluon/sn_gan/data.py  model.py  train.py  utils.py). I will modify these files and I don`t know if there are other files related to the 'example/gluon/dcgan.py' ?

**Comment by aaronmarkham** [comment]
> > Also, you're deleting a file that has links to it elsewhere in the repo, specifically /example/gluon/sn_gan. Make sure you deal with that. Delete the links or update as appropriate...
> 
> @aaronmarkham Thanks for your suggestions, I only found that there are relative link in the comment section of these files(/example/gluon/sn_gan/data.py model.py train.py utils.py). I will modify these files and I don`t know if there are other files related to the 'example/gluon/dcgan.py' ?

@pengxin99 I think your update covers it.

**Comment by aaronmarkham** [comment]
Once you fix https://github.com/dmlc/web-data/pull/127 I can merge that and then you can delete the png files from this PR.

**Comment by aaronmarkham** [comment]
Looks like you have to update the image links now that those images are merged in the web-data repo. Otherwise seems good to go!

**Comment by juliusshufan** [comment]
@aaronmarkham Appreciate for your prompt response, for "image links", do you refer to the ones in README.md? If yes, I think they already direct to dmlc/web-data? Thanks. @pengxin99

**Comment by aaronmarkham** [comment]
When I click View on the Readme it goes here: https://github.com/apache/incubator-mxnet/blob/1f0e7cbda9665d5e94c9de45d2317cae928a0948/example/gluon/DCGAN/README.md
And I see broken images.

File: example/gluon/DCGAN/README.md [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+This implementation uses [inception_score](https://github.com/openai/improved-gan) to evaluate the model.
+
+You can use this reference implementation on the MNIST and CIFAR-10 datasets.
+
+
+#### Generated image output examples from the CIFAR-10 dataset
+![Generated image output examples from the CIFAR-10 dataset](https://github.com/dmlc/web-data/tree/master/mxnet/example/gluon/DCGAN/fake_img_iter_13900.png)
+
+#### Generated image output examples from the MNIST dataset
+![Generated image output examples from the MNIST dataset](https://github.com/dmlc/web-data/tree/master/mxnet/example/gluon/DCGAN/fake_img_iter_21700.png)
```
Comment (aaronmarkham) [inline example/gluon/DCGAN/README.md]:
```suggestion
![Generated image output examples from the MNIST dataset](https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/example/gluon/DCGAN/fake_img_iter_21700.png)
```

File: example/gluon/DCGAN/README.md [hunk 21]
```diff
@@ -0,0 +1,52 @@
+# DCGAN in MXNet
+
+[Deep Convolutional Generative Adversarial Networks(DCGAN)](https://arxiv.org/abs/1511.06434) implementation with Apache MXNet GLUON.
+This implementation uses [inception_score](https://github.com/openai/improved-gan) to evaluate the model.
+
+You can use this reference implementation on the MNIST and CIFAR-10 datasets.
+
+
+#### Generated image output examples from the CIFAR-10 dataset
+![Generated image output examples from the CIFAR-10 dataset](https://github.com/dmlc/web-data/tree/master/mxnet/example/gluon/DCGAN/fake_img_iter_13900.png)
```
Comment (aaronmarkham) [inline example/gluon/DCGAN/README.md]:
```suggestion
![Generated image output examples from the CIFAR-10 dataset](https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/example/gluon/DCGAN/fake_img_iter_13900.png)
```

File: example/gluon/DCGAN/README.md [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+
+#### Generated image output examples from the MNIST dataset
+![Generated image output examples from the MNIST dataset](https://github.com/dmlc/web-data/tree/master/mxnet/example/gluon/DCGAN/fake_img_iter_21700.png)
+
+#### inception_score in cpu and gpu (the real image`s score is around 3.3)
+CPU & GPU
+
+![inception score with CPU](https://github.com/dmlc/web-data/tree/master/mxnet/example/gluon/DCGAN/inception_score_cifar10_cpu.png)
```
Comment (aaronmarkham) [inline example/gluon/DCGAN/README.md]:
```suggestion
![inception score with CPU](https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/example/gluon/DCGAN/inception_score_cifar10_cpu.png)
```

File: example/gluon/DCGAN/README.md [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+![inception score with GPU](https://github.com/dmlc/web-data/tree/master/mxnet/example/gluon/DCGAN/inception_score_cifar10.png)
```
Comment (aaronmarkham) [inline example/gluon/DCGAN/README.md]:
```suggestion
![inception score with GPU](https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/example/gluon/DCGAN/inception_score_cifar10.png)
```

**Comment by juliusshufan** [comment]
@aaronmarkham Thanks for the suggestions, I change it accordingly, and seems working.

**Comment by juliusshufan** [comment]
Thansk for your effort on review and the approval. @aaronmarkham 
@sandeep-krishnamurthy May I have your comments on next?
